File: src/execution/paper.ts
/**
 * Paper trading executor — simulates order fills at current market price.
 *
 * Simulates realistic execution by applying:
 *  - Slippage model: 0.05% on buys, 0.03% on sells (conservative estimate)
 *  - Per-account balance tracking (starts at $10,000 USDC by default)
 *  - Execution latency simulation (50-150ms random delay)
 *
 * All paper trades are logged identically to live trades, just with
 * paperTrading=true. The self-healing engine treats them the same.
 */

import { randomUUID } from 'crypto';
import { log } from '../storage/database.js';
import type { Trade } from '../types.js';

const INITIAL_BALANCE_USD = 10_000;
const SLIPPAGE_BUY  = 0.0005; // 0.05%
const SLIPPAGE_SELL = 0.0003; // 0.03%
const COMMISSION    = 0.006;  // 0.6% Coinbase Advanced maker fee (conservative)

let balance = INITIAL_BALANCE_USD;
const holdings = new Map<string, number>(); // symbol → quantity

function simulateDelay(): Promise<void> {
  const ms = 50 + Math.random() * 100;
  return new Promise(resolve => setTimeout(resolve, ms));
}

export async function paperBuy(
  symbol: string,
  productId: string,
  sizeUsd: number,
  positionId: string,
  marketPrice: number,
): Promise<Trade> {
  await simulateDelay();

  if (sizeUsd > balance) {
    sizeUsd = balance;
    log('warn', `Paper: insufficient balance, capped order to $${sizeUsd.toFixed(2)}`, { symbol });
  }

  const fillPrice = marketPrice * (1 + SLIPPAGE_BUY);
  const commission = sizeUsd * COMMISSION;
  const netSizeUsd = sizeUsd - commission;
  const quantity = netSizeUsd / fillPrice;

  balance -= sizeUsd;
  holdings.set(symbol, (holdings.get(symbol) ?? 0) + quantity);

  const trade: Trade = {
    id: randomUUID(),
    positionId,
    side: 'buy',
    symbol,
    quantity,
    sizeUsd,
    price: fillPrice,
    status: 'paper',
    paperTrading: true,
    placedAt: Date.now(),
  };

  log('trade', `[PAPER] BUY ${symbol} $${sizeUsd.toFixed(0)} @ ${fillPrice.toFixed(4)} (slip +${(SLIPPAGE_BUY * 100).toFixed(2)}% fee ${commission.toFixed(2)})`, {
    symbol, data: { fillPrice, quantity, commission, balanceAfter: balance },
  });

  return trade;
}

export async function paperSell(
  symbol: string,
  productId: string,
  quantity: number,
  positionId: string,
  marketPrice: number,
): Promise<Trade> {
  await simulateDelay();

  const held = holdings.get(symbol) ?? 0;
  const actualQty = Math.min(quantity, held);

  const fillPrice = marketPrice * (1 - SLIPPAGE_SELL);
  const grossProceeds = actualQty * fillPrice;
  const commission = grossProceeds * COMMISSION;
  const netProceeds = grossProceeds - commission;

  holdings.set(symbol, held - actualQty);
  balance += netProceeds;

  const trade: Trade = {
    id: randomUUID(),
    positionId,
    side: 'sell',
    symbol,
    quantity: actualQty,
    sizeUsd: netProceeds,
    price: fillPrice,
    status: 'paper',
    paperTrading: true,
    placedAt: Date.now(),
  };

  log('trade', `[PAPER] SELL ${symbol} ${actualQty.toFixed(6)} @ ${fillPrice.toFixed(4)} net $${netProceeds.toFixed(2)} (slip -${(SLIPPAGE_SELL * 100).toFixed(2)}%)`, {
    symbol, data: { fillPrice, quantity: actualQty, commission, balanceAfter: balance },
  });

  return trade;
}

export function getPaperBalance(): number { return balance; }
export function getPaperHoldings(): Map<string, number> { return new Map(holdings); }
export function resetPaperAccount(): void {
  balance = INITIAL_BALANCE_USD;
  holdings.clear();
}

File: src/execution/paper.py
"""Paper trading simulator."""

import random
import threading
import time
import uuid

from src.config import env
from src.storage.database import log
from src.types import Trade

_INITIAL_BALANCE = 10_000
_SLIPPAGE_BUY = 0.0005
_SLIPPAGE_SELL = 0.0003
_MIN_ORDER_USD = 1.0

_lock = threading.Lock()
_balance = _INITIAL_BALANCE
_holdings: dict[str, float] = {}


def _simulate_delay() -> None:
    time.sleep(0.05 + random.random() * 0.1)


def paper_buy(symbol: str, product_id: str, size_usd: float,
              position_id: str, market_price: float) -> Trade:
    global _balance
    _simulate_delay()

    # Price sanity check
    if market_price <= 0:
        log("error", f"Paper: invalid market_price={market_price} for {symbol}", symbol=symbol)
        return Trade(
            id=str(uuid.uuid4()), position_id=position_id, side="buy",
            symbol=symbol, quantity=0, size_usd=0,
            price=0, status="failed", paper_trading=True,
            placed_at=time.time() * 1000, error="Invalid market_price <= 0",
        )

    # Order size validation
    if size_usd < _MIN_ORDER_USD:
        log("error", f"Paper: order size ${size_usd:.2f} below minimum ${_MIN_ORDER_USD:.2f}", symbol=symbol)
        return Trade(
            id=str(uuid.uuid4()), position_id=position_id, side="buy",
            symbol=symbol, quantity=0, size_usd=0,
            price=0, status="failed", paper_trading=True,
            placed_at=time.time() * 1000, error=f"Order size below minimum ${_MIN_ORDER_USD}",
        )

    max_position_usd = env.max_position_usd
    if size_usd > max_position_usd:
        log("warn", f"Paper: order size ${size_usd:.2f} exceeds max ${max_position_usd:.2f}, capping",
            symbol=symbol)
        size_usd = max_position_usd

    with _lock:
        if size_usd > _balance:
            original_size = size_usd
            size_usd = _balance
            log("warn", f"Paper: insufficient balance, capped order from ${original_size:.2f} to ${size_usd:.2f}",
                symbol=symbol)

        # Negative balance protection
        if size_usd <= 0:
            log("error", f"Paper: zero balance, cannot buy {symbol}", symbol=symbol)
            return Trade(
                id=str(uuid.uuid4()), position_id=position_id, side="buy",
                symbol=symbol, quantity=0, size_usd=0,
                price=0, status="failed", paper_trading=True,
                placed_at=time.time() * 1000, error="Insufficient balance",
            )

        fill_price = market_price * (1 + _SLIPPAGE_BUY)
        commission = size_usd * env.commission_per_side
        net_size_usd = size_usd - commission
        quantity = net_size_usd / fill_price

        _balance -= size_usd
        _holdings[symbol] = _holdings.get(symbol, 0) + quantity
        balance_after = _balance

    log("trade",
        f"[PAPER] BUY {symbol} ${size_usd:.0f} @ {fill_price:.4f} (slip +{_SLIPPAGE_BUY*100:.2f}% fee {commission:.2f})",
        symbol=symbol,
        data={"fill_price": fill_price, "quantity": quantity, "commission": commission, "balance_after": balance_after})

    return Trade(
        id=str(uuid.uuid4()), position_id=position_id, side="buy",
        symbol=symbol, quantity=quantity, size_usd=size_usd,
        price=fill_price, status="paper", paper_trading=True,
        placed_at=time.time() * 1000, commission=commission,
    )


def paper_sell(symbol: str, product_id: str, quantity: float,
               position_id: str, market_price: float) -> Trade:
    global _balance
    _simulate_delay()

    # Price sanity check — division by zero guard
    if market_price <= 0:
        log("error", f"Paper: invalid market_price={market_price} for sell of {symbol}", symbol=symbol)
        return Trade(
            id=str(uuid.uuid4()), position_id=position_id, side="sell",
            symbol=symbol, quantity=0, size_usd=0,
            price=0, status="failed", paper_trading=True,
            placed_at=time.time() * 1000, error="Invalid market_price <= 0",
        )

    if quantity <= 0:
        log("error", f"Paper: invalid sell quantity={quantity} for {symbol}", symbol=symbol)
        return Trade(
            id=str(uuid.uuid4()), position_id=position_id, side="sell",
            symbol=symbol, quantity=0, size_usd=0,
            price=0, status="failed", paper_trading=True,
            placed_at=time.time() * 1000, error="Invalid quantity <= 0",
        )

    with _lock:
        held = _holdings.get(symbol, 0)
        actual_qty = min(quantity, held)

        # Log warning if sell request exceeds holdings
        if quantity > held:
            log("warn",
                f"Paper: sell request {quantity:.8f} exceeds holdings {held:.8f} for {symbol}, "
                f"selling actual held amount {actual_qty:.8f}",
                symbol=symbol,
                data={"requested_qty": quantity, "held_qty": held, "actual_qty": actual_qty})

        if actual_qty <= 0:
            log("error", f"Paper: no holdings of {symbol} to sell", symbol=symbol)
            return Trade(
                id=str(uuid.uuid4()), position_id=position_id, side="sell",
                symbol=symbol, quantity=0, size_usd=0,
                price=0, status="failed", paper_trading=True,
                placed_at=time.time() * 1000, error="No holdings to sell",
            )

        fill_price = market_price * (1 - _SLIPPAGE_SELL)

        # Division by zero guard on fill_price
        if fill_price <= 0:
            log("error", f"Paper: computed fill_price={fill_price} <= 0 for {symbol}", symbol=symbol)
            return Trade(
                id=str(uuid.uuid4()), position_id=position_id, side="sell",
                symbol=symbol, quantity=0, size_usd=0,
                price=0, status="failed", paper_trading=True,
                placed_at=time.time() * 1000, error="Fill price <= 0",
            )

        gross_proceeds = actual_qty * fill_price
        commission = gross_proceeds * env.commission_per_side
        net_proceeds = gross_proceeds - commission

        _holdings[symbol] = held - actual_qty
        # Negative balance protection: only add non-negative proceeds
        if net_proceeds > 0:
            _balance += net_proceeds
        balance_after = _balance

    log("trade",
        f"[PAPER] SELL {symbol} {actual_qty:.6f} @ {fill_price:.4f} net ${net_proceeds:.2f} (slip -{_SLIPPAGE_SELL*100:.2f}%)",
        symbol=symbol,
        data={"fill_price": fill_price, "quantity": actual_qty, "commission": commission, "balance_after": balance_after})

    return Trade(
        id=str(uuid.uuid4()), position_id=position_id, side="sell",
        symbol=symbol, quantity=actual_qty, size_usd=net_proceeds,
        price=fill_price, status="paper", paper_trading=True,
        placed_at=time.time() * 1000, commission=commission,
    )


def get_paper_balance() -> float:
    with _lock:
        return _balance


def get_paper_holdings() -> dict[str, float]:
    with _lock:
        return dict(_holdings)


def reset_paper_account() -> None:
    global _balance
    with _lock:
        _balance = _INITIAL_BALANCE
        _holdings.clear()
